tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/LegacyPopTest.cs

```diff
@@ -0,0 +1,185 @@
+            at = (cca.AppTokenCache as ITokenCacheInternal).Accessor.GetAllAccessTokens().Single();
+            Assert.AreEqual(at.KeyId, _currentPopKey.KeyId);
+
+            _currentPopKey = CreateRsaSecurityKey();
+
+            result = await cca.AcquireTokenForClient(settings.AppScopes)
+                .WithClientAssertion(
+                    (tokenEndpoint, ct) =>
+                    Task.FromResult(CreateMs10ATPOPBodyParams(settings.ClientId, tokenEndpoint, clientCredsCert, _currentPopKey, true)))
+                .WithKeyId(_currentPopKey.KeyId)
+             .ExecuteAsync()
+             .ConfigureAwait(false);
+
+            Assert.AreEqual(TokenSource.IdentityProvider, result.AuthenticationResultMetadata.TokenSource);
+            at = (cca.AppTokenCache as ITokenCacheInternal).Accessor.GetAllAccessTokens().Single();
+            Assert.AreEqual(at.KeyId, _currentPopKey.KeyId);
+        }
+
+        private static RsaSecurityKey CreateRsaSecurityKey()
+        {
+#if NET472
+            RSA rsa = RSA.Create(2048);
+#else
+            RSA rsa = new RSACryptoServiceProvider(2048);
+#endif
+            // the reason for creating the RsaSecurityKey from RSAParameters is so that a SignatureProvider created with this key
+            // will own the RSA object and dispose it. If we pass a RSA object, the SignatureProvider does not own the object, the RSA object will not be disposed.
+            RSAParameters rsaParameters = rsa.ExportParameters(true);
+            RsaSecurityKey rsaSecuirtyKey = new RsaSecurityKey(rsaParameters) { KeyId = CreateRsaKeyId(rsaParameters) };
+            rsa.Dispose();
+            return rsaSecuirtyKey;
+        }
+
+        private static string CreateRsaKeyId(RSAParameters rsaParameters)
+        {
+            byte[] kidBytes = new byte[rsaParameters.Exponent.Length + rsaParameters.Modulus.Length];
+            Array.Copy(rsaParameters.Exponent, 0, kidBytes, 0, rsaParameters.Exponent.Length);
+            Array.Copy(rsaParameters.Modulus, 0, kidBytes, rsaParameters.Exponent.Length, rsaParameters.Modulus.Length);
+            using (var sha2 = SHA256.Create())
+                return Base64UrlEncoder.Encode(sha2.ComputeHash(kidBytes));
+        }
+
+        private static IReadOnlyList<KeyValuePair<string, string>> CreateMs10ATPOPBodyParams(
+           string clientId,
+           string audience,
+           X509Certificate2 clientCredCert,
+           RsaSecurityKey popKey,
+           bool includeX5cClaim)
+        {
+            var clientCredsSigningCredentials = new SigningCredentials(new X509SecurityKey(clientCredCert), SecurityAlgorithms.RsaSha256, SecurityAlgorithms.Sha256);
+            string request = CreateMs10ATPOPAssertion(clientId, audience, clientCredsSigningCredentials, popKey, includeX5cClaim);
+            return new List<KeyValuePair<string, string>>(1) { new KeyValuePair<string, string>("request", request) };
```

Comment:
So for Ms10ATPOP there has to be a body param request? Does this func need to return a collection of key-values or are you making this a bit more generic to apply to more different cases?

Reply:
Making it generic. I am going to use this to send normal client credentials as well. It helps with having a singleton CCA.